[Blog]  Top 5 Things andAzure Developer Needs to Know: Kubernetes Infrastructure (#505)
pulumi/pulumi-hugo: themes/default/content/blog/top-5-things-for-azure-devs-kubernetes-infrastructure/index.md

<!--more-->

## A Kubernetes Review

If you're not familiar with Kubernetes concepts and terminology, the Getting Started with Kubernetes series can help get you up to speed.

- [Getting Started With Kubernetes: Clusters and Nodes]({{< relref "/blog/getting-started-with-k8s-part1" >}})
- [Getting Started With Kubernetes: Application Basics]({{< relref "/blog/getting-started-with-k8s-part2/" >}})
- [Getting Started With Kubernetes: Advanced Deployment]({{< relref "/blog/getting-started-with-k8s-part3" >}})
- [Getting Started with Kubernetes: Stateful Applications]({{< relref "/blog/getting-started-with-k8s-part4" >}})
- [Getting Started With Kubernetes: Networking]({{< relref "/blog/getting-started-with-k8s-part5" >}})
- [Getting Started With Kubernetes: Day 2]({{< relref "/blog/getting-started-with-k8s-part6" >}})

## Azure Kubernetes Service

Azure Kubernetes Service (AKS) is a hosted Kubernetes service. Azure manages the Kubernetes master nodes, and you are responsible for managing the agent or worker nodes. You only pay for the worker nodes in your cluster that make up your application.

> Kubernetes nodes are the worker machines that can be either physical or virtual. AKS nodes use Azure virtual machines (VMs), and you can add storage, upgrade cluster components, or  even run multiple node pools with mixed operating systems.

Reviewer: ```suggestion
Kubernetes nodes are the worker machines that can be either physical or virtual. AKS nodes use Azure virtual machines (VMs); and you can add storage, upgrade cluster components, or even run multiple node pools with mixed operating systems.
```

> You can create an AKS cluster with:

Reviewer: ```suggestion
You can create an AKS cluster with
```

- the Azure CLI,
- the Azure portal,
- PowerShell, and
- templates, such as Azure Resource Manager (ARM) templates.

### AKS features

> Configuring and deploying Kubernetes can be complex. AKS provides manyfeatures to simplify the process. They include:

Reviewer: ```suggestion
Configuring and deploying Kubernetes can be complex. AKS provides many features to simplify the process, including the following features:
```

- Cluster node and pod scaling adjust the number of nodes or pods as resource demand changes. [Pods can scale horizontally, and the cluster can also autoscale](https://docs.microsoft.com/en-us/azure/aks/tutorial-kubernetes-scale?tabs=azure-cli).
- Kubernetes is a fast-moving open source project, and AKS supports multiple versions. You can [upgrade the cluster nodes](https://docs.microsoft.com/en-us/azure/aks/node-image-upgrade) without interrupting the application.
- If you perform machine learning, AKS can create [GPU-enabled node pools](https://docs.microsoft.com/en-us/azure/aks/gpu-cluster) for compute-intensive, graphics-intensive, and visualization workloads.
- AKS supports access control through [Kubernetes RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) and Azure [Active Directory](https://docs.microsoft.com/en-us/azure/aks/managed-aad).
- Integrated logging and monitoring with [Azure Monitor for Containers](https://docs.microsoft.com/en-us/azure/azure-monitor/containers/container-insights-overview) collects memory and processor metrics from containers, nodes, and controllers.

> - AKS lets you mount static or dynamic storage volumes for persistent data. Storage volumes use Azure Disks for single pod access or Azure Files for concurrent pod access.

Reviewer: Should there be any links here?

Author: ```suggestion
- AKS lets you mount [static or dynamic storage volumes](https://docs.microsoft.com/en-us/azure/storage/common/storage-introduction) for persistent data. Storage volumes use Azure Disks for single pod access or Azure Files for concurrent pod access.
```

> - [Virtual networking](https://docs.microsoft.com/en-us/azure/aks/concepts-network) for pods in the cluster and other nodes using a peered virtual network. The network also provides ingress for applications with HTTP application routing.

Reviewer: ```suggestion
- [Virtual networking](https://docs.microsoft.com/en-us/azure/aks/concepts-network) for pods in the cluster and other nodes let you use a peered virtual network. The network also provides ingress for applications with HTTP application routing.
```

> - Integration with development tools for Helm and a [Visual Studio Code extension for Kubernetes](https://marketplace.visualstudio.com/items?itemName=ms-kubernetes-tools.vscode-kubernetes-tools).

Reviewer: ```suggestion
- Integration with development tools for Helm and a [Visual Studio Code extension for Kubernetes](https://marketplace.visualstudio.com/items?itemName=ms-kubernetes-tools.vscode-kubernetes-tools) improves the developer experience.
```

> - Support for Docker images, including storage with [Azure Container Registry](https://azure.microsoft.com/en-us/services/container-registry/).

Reviewer: ```suggestion
- Support for Docker images, including storage with [Azure Container Registry](https://azure.microsoft.com/en-us/services/container-registry/), makes deployment easier.
```

## Create an AKS Cluster with the Azure Portal

We'll use the Azure Portal to illustrate the steps to configure and deploy an AKS Cluster.

**Step 1**: Create a Kubernetes Service.

> - Open the [Azure Portal](https:portal.azure.com) and select `Create Resource`.

Reviewer: ```suggestion
- Open the [Azure Portal](https:portal.azure.com) and select **Create Resource**.
```

> - Select `Containers` > `Kubernetes Service`

Reviewer: ```suggestion
- Select **Containers > Kubernetes Service**.
```

> - Select `Create`

Reviewer: ```suggestion
- Select **Create**.
```

![Create Kubernetes service](./kubernetes_service.png)

**Step 2**: Create an AKS cluster.

> For this example, we will configure several options in the `Basic` window but use defaults for most options.

Reviewer: ```suggestion
For this example, we will configure several options in the **Basic** window but use defaults for most options.
```

> - In `Subscription`, Select an `Azure Subscription`

Reviewer: ```suggestion
- In **Subscription**, select an **Azure Subscription**.
```

> - Create or select a [Resource Group](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/overview)

Reviewer: ```suggestion
- Create or select a [Resource Group](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/overview).
```

> - In `Cluster details`, set the `Kubernetes cluster name`

Reviewer: ```suggestion
- In **Cluster details**, set the **Kubernetes cluster name**.
```

> - Change the `Region` if needed

Reviewer: ```suggestion
- Change the **Region** if needed.
```

> Use the default values for `Primary node pool` and select `Next: node pools`.

Reviewer: ```suggestion
Use the default values for **Primary node pool**, and select **Next: node pools**.
```

![Create AKS Cluster](./create_aks_cluster.png)

**Step 3**: Configure node pools.

A node pool is a logical grouping of nodes. Nodes in a pool can have different virtual machines, different Kubernetes versions, and other attributes. You can use different node pools for different purposes, such as grouping workflows, e.g., a node pool for production and one for dev or test.

For this example, use the default values for node pools.

![Node pools](./node_pools.png)

**Step 4**: Configure authentication.

> You can authenticate, authorize, secure, and control access to Kubernetes clusters with:

Reviewer: ```suggestion
You can authenticate, authorize, secure, and control access to Kubernetes clusters with
```

> - [Kubernetes role-based access control](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) (Kubernetes RBAC) to grant access to service accounts

Reviewer: ```suggestion
- [Kubernetes role-based access control](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) (Kubernetes RBAC) to grant access to service accounts or
```

- [Azure Active Directory integration](https://docs.microsoft.com/en-us/azure/aks/managed-aad) that provides a multi-tenant, cloud-based directory and identity management service.

> To configure `Authentication` for this example, set the `Authentication method` to `System-assigned managed identity`. You can set `Role-based access control (RBAC)` to enabled, but it's unnecessary for the example. Use the default value for the `Node pool OS disk encryption` type.

Reviewer: ```suggestion
To configure **Authentication** for this example, set the **Authentication method** to **System-assigned managed identity**. You can set **Role-based access control (RBAC)** to enabled, but it's unnecessary for the example. Use the default value for the **Node pool OS disk encryption** type.
```

![Authentication](./authentication.png)

**Step 5**: Configure networking

> AKS can use either `kubenet` or `Azure CNI` networking. Kubenet is the default configuration for AKS cluster creation.

Reviewer: Note: I'm not changing these to bold because, while they are in the screenshot, they're actual products/things. I think it makes more sense to leave them as is since the text isn't explicitly referring to the UI.

With `kubenet`, Azure creates and configures the virtual network. However, only the nodes receive a routable IP address and pods use a NAT to communicate with resources outside the AKS cluster. This approach reduces the number of IP addresses you need to reserve in your network space for pods to use.

`Azure CNI`assigns an IP address to every pod, which makes them directly accessible. The IP addresses must be unique and  planned in advance. Each node must be configured for the maximum number of pods to reserve IP addresses per node. Without accounting for the number of pods, the network can run out of IP addresses to allocate or necessitate the need to rebuild clusters with a larger subnet.

For this example, we'll use `kubenet` for it's simplicity, but in production `Azure CNI` maybe the better choice for applications.

![Networking](./networking.png)

**Step 6**: Complete the deployment

> Select `Review + create` to deploy. Deployment takes several minutes to complete. On completion, you can verify the `Deployment details` or select `Connect to cluster`.

Reviewer: ```suggestion
Select **Review + create** to deploy. Deployment takes several minutes to complete. On completion, you can verify the **Deployment details** or select **Connect to cluster**.
```

![Deployment complete](deployment_complete.png)

**Step 7**: Connect to the cluster

> You can manage your cluster with [`kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl), a command line tool for managing Kubernetes. You can install kubectl on [linux](https://kubernetes.io/docs/tasks/tools/install-kubectl-linux), [macOS](https://kubernetes.io/docs/tasks/tools/install-kubectl-macos), and [Windows](https://kubernetes.io/docs/tasks/tools/install-kubectl-windows).

Reviewer: ```suggestion
You can manage your cluster with [`kubectl`](https://kubernetes.io/docs/tasks/tools/#kubectl), a command line tool for managing Kubernetes. You can install `kubectl` on [linux](https://kubernetes.io/docs/tasks/tools/install-kubectl-linux), [macOS](https://kubernetes.io/docs/tasks/tools/install-kubectl-macos), and [Windows](https://kubernetes.io/docs/tasks/tools/install-kubectl-windows).
```

    {{% /choosable %}}
    {{% choosable language "typescript" %}}

        // Create an Azure Resource Group
        const resourceGroup = new resources.ResourceGroup("azure-go-aks");

        // Create an AD service principal
        const adApp = new azuread.Application("aks", {
            displayName: "aks",
        });
        const adSp = new azuread.ServicePrincipal("aksSp", {
            applicationId: adApp.applicationId,
        });

        // Generate random password
        const password = new random.RandomPassword("password", {
            length: 20,
            special: true,
        });

        // Create the Service Principal Password
        const adSpPassword = new azuread.ServicePrincipalPassword("aksSpPassword", {
            servicePrincipalId: adSp.id,
            value: password.result,
            endDate: "2099-01-01T00:00:00Z",
        });

        // Generate an SSH key
        const sshKey = new tls.PrivateKey("ssh-key", {
            algorithm: "RSA",
            rsaBits: 4096,
        });

    {{% /choosable %}}
{{< /chooser >}}

**Step 2**: Configure the AKS cluster

We set the configuration options in `ManagedClusterAgentPoolProfileArgs`:

> - Count: Number of virtual machines
> - MaxPods: Maximum number of pods that can run on a node.
> - Mode: Sets the type of pool node, which can be [system or user](https://docs.microsoft.com/en-us/azure/aks/use-system-pools)
> - Name: Name of the pool node
> - OsType: Specifies the OS type
> - Type: Choose between a [`VirtualMachineScaleSet`](https://docs.microsoft.com/en-us/azure/aks/use-system-pools) or an [`AvailabilitySet`](https://docs.microsoft.com/en-us/azure/aks/availability-zones)
> - VmSize: The virtual machine size used by the cluster.

Reviewer: ```suggestion
- `Count`: Number of virtual machines
- `MaxPods`: Maximum number of pods that can run on a node.
- `Mode`: Sets the type of pool node, which can be [system or user](https://docs.microsoft.com/en-us/azure/aks/use-system-pools)
- `Name`: Name of the pool node
- `OsType`: Specifies the OS type
- `Type`: Choose between a [`VirtualMachineScaleSet`](https://docs.microsoft.com/en-us/azure/aks/use-system-pools) or an [`AvailabilitySet`](https://docs.microsoft.com/en-us/azure/aks/availability-zones)
- `VmSize`: The virtual machine size used by the cluster.
```

			})
			if err != nil {
				return "", err
			}
			encoded := creds.Kubeconfigs[0].Value
			kubeconfig, err := base64.StdEncoding.DecodeString(encoded)
			if err != nil {
				return "", err
			}
			return string(kubeconfig), nil
		}))

    {{% /choosable %}}
    {{% choosable language "python" %}}

        # Export kubeconfig
        encoded = creds.kubeconfigs[0].value
        kubeconfig = encoded.apply(
            lambda enc: base64.b64decode(enc).decode())
        pulumi.export("kubeconfig", kubeconfig)

        {{% /choosable %}}
        {{% choosable language "typescript" %}}

    {{% /choosable %}}
    {{% choosable language "typescript" %}}

        const creds = pulumi.all([cluster.name, resourceGroup.name]).apply(([clusterName, rgName]) => {
            return containerservice.listManagedClusterUserCredentials({
                resourceGroupName: rgName,
                resourceName: clusterName,
            });
        });

        const encoded = creds.kubeconfigs[0].value;
        export const kubeconfig = encoded.apply(enc => Buffer.from(enc, "base64").toString());

    {{% /choosable %}}
{{< /chooser >}}

**Step 4**: Connect to the AKS cluster

> We can use the `kubeconfig` file connect to the cluster. It's common practice to copy the `kubeconfig` file to `~/.kube/config` which is the directory that `kubectl` looks for the file. However, you can use any directory with the `--kubeconfig` flag, e.g.,

Reviewer: ```suggestion
We can use the `kubeconfig` file connect to the cluster. It's common practice to copy the `kubeconfig` file to `~/.kube/config`, which is the directory that `kubectl` looks for the file. However, you can use any directory with the `--kubeconfig` flag, e.g.,
```

```bash
$ kubectl --kubeconfig /path/to/kubeconfig_file get pods
```

> As this example shows, deploying an AKS cluster requires setting parameters. This can be done in the portal or with code. The advantage of code is that clusters can be created on demand without having to use the Azure portal. This is convenient when you have different environments, such as a dev/test environment and a production environment.

Reviewer: ```suggestion
As this example shows, deploying an AKS cluster requires setting parameters. This action can be done in the portal or with code. The advantage of code is that clusters can be created on demand without having to use the Azure portal. This setup is convenient when you have different environments, such as a dev/test environment and a production environment.
```

The complete code for deploying AKS is available on GitHub for [Typescript](https://github.com/pulumi/examples/tree/master/azure-ts-aks), [Python](https://github.com/pulumi/examples/tree/master/azure-py-aks), [Go](https://github.com/pulumi/examples/tree/master/azure-go-aks), and [C#](https://github.com/pulumi/examples/tree/master/azure-cs-aks).

## Summary

> Azure Kubernetes Service let's you deploy a Kubernetes cluster quickly and efficiently. An AKS deployment configures the worker nodes since the master nodes are provided by Azure. The Azure Portal provides a simplified interface for configuration, but if you want fine-grain control over the deployment, infrastructure as code is an option. This is particularly true for production deployments where the configuration requires tuning for an application deployment.

Reviewer: ```suggestion
Azure Kubernetes Service lets you deploy a Kubernetes cluster quickly and efficiently. An AKS deployment configures the worker nodes since the master nodes are provided by Azure. The Azure Portal provides a simplified interface for configuration, but if you want fine-grain control over the deployment, infrastructure as code is an option. This is particularly true for production deployments where the configuration requires tuning for an application deployment.
```
